storage/innobase/trx/trx0rseg.cc

```diff
@@ -187,16 +187,29 @@ bool trx_rseg_read_wsrep_checkpoint(const trx_rsegf_t* rseg_header, XID& xid)
 		return false;
 	}
 
-	xid.formatID = formatID;
-	xid.gtrid_length = static_cast<int>(
+	XID temp_xid;
+	temp_xid.formatID = formatID;
+	temp_xid.gtrid_length = static_cast<int>(
 		mach_read_from_4(
 			TRX_RSEG_WSREP_XID_GTRID_LEN + rseg_header));
 
-	xid.bqual_length = static_cast<int>(
+	temp_xid.bqual_length = static_cast<int>(
 		mach_read_from_4(
 			TRX_RSEG_WSREP_XID_BQUAL_LEN + rseg_header));
 
-	memcpy(xid.data, TRX_RSEG_WSREP_XID_DATA + rseg_header, XIDDATASIZE);
+	memcpy(temp_xid.data, TRX_RSEG_WSREP_XID_DATA + rseg_header, XIDDATASIZE);
+
+	long long tmp_seqno = wsrep_xid_seqno(&temp_xid);
+
+	/* Set new xid only if its seqno is larger than current
+	   seqno. */
+	if (tmp_seqno > (long long)wsrep_xid_seqno(&xid))
```

Comment:
Cast to `long long` should not be necessary. The return value of `wsrep_xid_seqno()` is `long long`.

Should we also check if `xid` is wsrep XID? e.g.

```if (!wsrep_is_wsrep_xid(&xid) || tmp_seqno > wsrep_xid_seqno(&xid))```

Reply:
In my opinion, this function should just read and return whatever xid it finds in rollback segment header. The comparison to some other xid, or any other logic, should be left to the caller.